libmultilabel/linear/metrics.py

```diff
@@ -4,12 +4,79 @@
 
 import numpy as np
 
+from ..common_utils import argsort_top_k
+
 __all__ = ['get_metrics',
            'compute_metrics',
            'tabulate_metrics',
            'MetricCollection']
 
 
+def _DCG(
+        preds: np.ndarray,
+        target: np.ndarray,
+        k: int = 5
+) -> np.ndarray:
+    """
+    Calculate the discounted cumulative gains (DCG).
+    """
+    k = min(preds.shape[-1], k)
+
+    # find the indices of the top-k predicted values in descending order
+    order = argsort_top_k(preds, k)
+    # also known as target sorted by the top-k preds
+    gains = np.take_along_axis(target, order, -1)
+
+    # the discount factor
+    discount = 1 / (np.log2(np.arange(gains.shape[1]) + 2))
+
+    # get the sum product over the last axis of gains and discount.
+    dcg = gains.dot(discount)
+
+    return dcg
+
+
+class NDCG:
+    def __init__(
+            self,
+            top_k: int
+    ) -> None:
+        """
+        Calculate the normalized DCG (nDCG).
+        Args:
+            top_k: consider only the top k elements for each instance.
+        """
+        self.top_k = top_k
+        self.score = 0
+        self.num_sample = 0
+
+    def update(
+            self,
+            preds: np.ndarray,
+            target: np.ndarray
+    ) -> None:
+        assert preds.shape == target.shape  # (batch_size, num_classes)
+
+        # the vanilla DCG
+        dcgs = _DCG(preds, target, self.top_k)
+        # the maximum possible DCG, also called Ideal DCG
+        idcgs = _DCG(target, target, self.top_k)
+        # the normalized DCG
+        ndcg_score = dcgs / idcgs
+
+        # deal with instances with all 0 labels and add up the results
```

Comment:
please check the situation of other metrics (also those for NN) and discuss in the meeting